### PR TITLE
[FLINK-8162] [kinesis-connector] Emit Kinesis' millisBehindLatest metric

### PR DESCRIPTION
**Pull request by casidiablo**
## What is the purpose of the change

- Emits [Kinesis' millisBehindLatest](http://docs.aws.amazon.com/kinesis/latest/APIReference/API_GetRecords.html) metric, which can be used to detect delays in the pipeline

## Brief change log

  - Publish `millisBehindLatest` as a gauge metric under the `Kinesis` group using `<shard_id>` as parameter
  - Updated metrics documentation

## Verifying this change

This change is already covered by existing tests, such as `ShardConsumerTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? `metrics.md` file

### CODE REVIEW

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 1]
```diff
@@ -96,6 +100,14 @@ protected ShardConsumer(KinesisDataFetcher<T> fetcherRef,
 							SequenceNumber lastSequenceNum,
 							KinesisProxyInterface kinesis) {
 		this.fetcherRef = checkNotNull(fetcherRef);
+		MetricGroup kinesisMetricGroup = fetcherRef.getRuntimeContext()
+			.getMetricGroup()
+			.addGroup("Kinesis");
+		kinesisMetricGroup
```
Comment (zentol) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: use `addGroup("shard_id", subscribedShard.getShard().getShardId())` instead and register the metric on the returned group.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 2]
```diff
@@ -96,6 +100,14 @@ protected ShardConsumer(KinesisDataFetcher<T> fetcherRef,
...
+			.getAllVariables()
+			.put("<shard_id>", subscribedShard.getShard().getShardId());
+
+		kinesisMetricGroup.gauge("millisBehindLatest", (Gauge<Long>) () -> millisBehindLatest);
```
Comment (zentol) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: the cast shouldn't be necessary

File: docs/monitoring/metrics.md [hunk 3]
```diff
@@ -1293,6 +1293,29 @@ Thus, in order to infer the metric identifier:
...
+      <th class="text-left" style="width: 26%">Metrics</th>
+      <th class="text-left" style="width: 48%">Description</th>
+      <th class="text-left" style="width: 8%">Type</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th rowspan="1">Operator</th>
+      <td>millisBehindLatest</td>
+      <td>The number of milliseconds the <a>GetRecords</a> response is from the tip of the stream,
```
Comment (tzulitai) [inline docs/monitoring/metrics.md]: Just a matter of preference here: I prefer the term "head of the stream" instead of tip.
You can ignore this if you disagree.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 4]
```diff
@@ -96,6 +99,15 @@ protected ShardConsumer(KinesisDataFetcher<T> fetcherRef,
 							SequenceNumber lastSequenceNum,
 							KinesisProxyInterface kinesis) {
 		this.fetcherRef = checkNotNull(fetcherRef);
+		MetricGroup kinesisMetricGroup = fetcherRef.getRuntimeContext()
+			.getMetricGroup()
+			.addGroup("Kinesis")
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: Do we really need this group? The metric is already bounded to the current subtask, which should provide enough context that it is Kinesis-related since we're the Kinesis consumer, no?

Comment (tzulitai) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: Would be best if @zentol also comments on this.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 5]
```diff
@@ -96,6 +99,15 @@ protected ShardConsumer(KinesisDataFetcher<T> fetcherRef,
 							SequenceNumber lastSequenceNum,
 							KinesisProxyInterface kinesis) {
 		this.fetcherRef = checkNotNull(fetcherRef);
+		MetricGroup kinesisMetricGroup = fetcherRef.getRuntimeContext()
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: I feel that passing the `StreamingRuntimeContext` all the way here just to register metrics, is not a good idea.
Is it possible we register the metrics in `FlinkKinesisConsumer` instead? That also makes it more visible what metrics the consumer exposes without having to dig all the way to this internal `ShardConsumer` thread.

File: docs/monitoring/metrics.md (hunk 3 above)
Comment (casidiablo) [reply docs/monitoring/metrics.md]: I'm OK changing it. That's actually just a copy&past from Amazon docs.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java (hunk 5 above)
Comment (casidiablo) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: I think we can't register the metric in `FlinkKinesisConsumer`, since we need it to be associated with a particular shard id. But I could do it from the `KinesisDataFetcher` instead, which already has access to the runtime context.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 6]
```diff
@@ -70,35 +71,45 @@
...
-							KinesisProxyInterface kinesis) {
+							KinesisProxyInterface kinesis,
+							MetricGroup kinesisMetricGroup) {
 		this.fetcherRef = checkNotNull(fetcherRef);
 		this.subscribedShardStateIndex = checkNotNull(subscribedShardStateIndex);
 		this.subscribedShard = checkNotNull(subscribedShard);
 		this.lastSequenceNum = checkNotNull(lastSequenceNum);
+
+		checkNotNull(kinesisMetricGroup)
+			.gauge("millisBehindLatest", () -> millisBehindLatest);
```
Comment (casidiablo) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: Will this work? Am I allowed to use lambdas? What Java version has to be supported?

Comment (casidiablo) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: Bump 👊

Comment (zentol) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: yes you can use lambdas.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java [hunk 7]
```diff
@@ -542,6 +545,16 @@ public int registerNewSubscribedShardState(KinesisStreamShardState newSubscribed
...
 	}
 
+	/**
+	 * Registers a metric group associated with the shard id of the provided {@link KinesisStreamShardState shardState}.
+	 */
+	private MetricGroup buildMetricGroupForShard(KinesisStreamShardState shardState) {
+		return runtimeContext
+			.getMetricGroup()
+			.addGroup("Kinesis")
+			.addGroup("shard_id", shardState.getStreamShardHandle().getShard().getShardId());
```
Comment (zentol) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java]: Is the underscore important? Would prefer shardId for consistency.

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/groups/UnregisteredMetricsGroup.java [hunk 8]
```diff
@@ -107,7 +107,7 @@ public MetricGroup addGroup(String key, String value) {
 
 	@Override
 	public Map<String, String> getAllVariables() {
-		return Collections.emptyMap();
+		return new HashMap<>();
```
Comment (zentol) [inline flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/groups/UnregisteredMetricsGroup.java]: is this change necessary?

Comment (casidiablo) [reply flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/groups/UnregisteredMetricsGroup.java]: I think it was before (tests were failing because I was using `...getAllVariables().put("shard_id")...`). I will revert and test.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java [hunk 9]
```diff
@@ -542,6 +545,16 @@ public int registerNewSubscribedShardState(KinesisStreamShardState newSubscribed
 		}
 	}
 
+	/**
+	 * Registers a metric group associated with the shard id of the provided {@link KinesisStreamShardState shardState}.
+	 */
+	private MetricGroup buildMetricGroupForShard(KinesisStreamShardState shardState) {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java]: personal preference here: would prefer this to be named `registerMetricGroupForShard`.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java [hunk 10]
```diff
@@ -542,6 +545,16 @@ public int registerNewSubscribedShardState(KinesisStreamShardState newSubscribed
...
+		return runtimeContext
+			.getMetricGroup()
+			.addGroup("Kinesis")
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java]: is this necessary?

Comment (casidiablo) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java]: I think so. Just following the convention of other components (e.g. [kafka consumer](https://github.com/apache/flink/blob/00ad0eb120026817c79a5fed2c71dd6aa10e9ba6/flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java#L101))

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java [hunk 11]
```diff
@@ -542,6 +545,16 @@ public int registerNewSubscribedShardState(KinesisStreamShardState newSubscribed
...
 	}
 
+	/**
+	 * Registers a metric group associated with the shard id of the provided {@link KinesisStreamShardState shardState}.
+	 */
+	private MetricGroup registerMetricGroupForShard(KinesisStreamShardState shardState) {
+		return runtimeContext
+			.getMetricGroup()
+			.addGroup("Kinesis")
+			.addGroup("shardId", shardState.getStreamShardHandle().getShard().getShardId());
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java]: We probably should also add a `stream` group before `shardId`, since the Kinesis consumer allows subscribing to multiple streams.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java (hunk 10 above)
Comment (tzulitai) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java]: Ah I see, that makes sense then.

Could you then name it as `KinesisConsumer`, to be more consistent?

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java (hunk 11 above)
Comment (casidiablo) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java]: Great idea.

**Comment by tzulitai** [comment]
Thanks @casidiablo. Merging ...

**Comment by tzulitai** [comment]
Hi @casidiablo, just before merging, I made some refactoring of your PR to move the metrics reporting business out of the `ShardConsumer`: https://github.com/apache/flink/commit/14b01be8ec1c00b867aea8dc758c23fe247e05f4.

Let me know if you have any objections with that, if not I'll proceed to merge soon.
